Add Modal Block

Soare-Robert-Daniel opened the pull request:
<!-- Issues that this pull request closes. -->
Closes https://github.com/Codeinwp/otter-internals/issues/175
<!-- Should look like this: `Closes #1, #2, #3.` . -->

### Summary
<!-- Please describe the changes you made. -->

The same features like Popup except for Popup Settings tabs in the Inspector

#### Dynamic Overriding 

The `save` function of the modal saves only the inner content. The complete HTML structure is done via PHP, which overrides the default saving data. 

This method ensures that the modal is working under a valid license while keeping the original data safe. 

### Screenshots <!-- if applicable -->

https://github.com/user-attachments/assets/23ff20c7-f2df-4a55-b9f3-47d7288f063e

----

### Test instructions
<!-- Describe how this pull request can be tested. -->

> [!NOTE]
> This is PRO block, you will need Otter Pro

Modal is a popup with only the anchor option.

1. Insert a button and set an anchor
2. Insert a modal with the anche for the button
3. Check if it works like in the video.

<!--
#### Query
```javascript
new QueryQA().select('blocks').run()
```
-->

---- 

### Checklist before the final review

- [x] Included E2E or unit tests for the changes in this PR.
- [x] Visual elements are not affected by independent changes.
- [x] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [x] It loads additional script in frontend only if it is required.
- [x] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [x] In case of deprecation, old blocks are safely migrated.
- [x] It is usable in Widgets and FSE.
- [x] Copy/Paste is working if the attributes are modified.
- [x] PR is following [the best practices]()

Soare-Robert-Daniel commented:
@ineagu @HardeepAsrani When a Free user uses the Modal block which Pro, it will see this:

![image](https://github.com/user-attachments/assets/b939400d-9d4f-4897-9a05-b6323403b4e3)

Two questions:
- Is the upsell message good?
- Should we use a different icon for the block to be different from the popup (the popup icon is used in the image)?

Soare-Robert-Daniel commented on src/blocks/blocks/modal/block.json:
```diff
@@ -0,0 +1,150 @@
+		}
+	},
+	"editorStyle": "otter-popup-editor",
+	"style": "file:../popup/style.css",
```
@HardeepAsrani, I have a problem. Ideally, we should use the same CSS file for the popup since it is the same.

However, using the same CSS hook handler as the one in the popup does not seem to work. My workaround is just creating another entry, so we still have duplicates. 

Any tips on this?

HardeepAsrani replied:
This might be a stupid idea, but have you tried to register the same script twice with a different handle and see if that works? And does that add the same script twice or avoids doing that?

ineagu commented:
thanks for the video @Soare-Robert-Daniel . 

Can we have default content in the modal popup as we have in the popup block? 

Can we add the button as well when we add the modal block with the anchor configured both in the modal & button settings?

Or

Can we simply add a button once the modal block is added and (edit modal) would be a buton moved into the sidebar settings of the "special" button block? Along with the modal settings?

Which one seems most logical and easy to do for you? 

My concerns: 

I find it that is unnecessary to require 3 steps that the user might not find.
The edit modal button is disrupting the editor view design and it does't look the same as frontend.

HardeepAsrani commented:
@Soare-Robert-Daniel We also need to list it here: https://github.com/Codeinwp/otter-internals/issues/224

HardeepAsrani commented:
@Soare-Robert-Daniel, I think there has been some confusion about the approach here. As @ineagu mentioned in the issue, the Block should add the Popup and the Button, and not adding an extra step for the user to setup the button manually.

Also, might make sense to keep using JS for the frontend to avoid issues as this will contain multiple instances of inner blocks, one being for button and another being for the popup content.

Let me know if @ineagu what you think.

HardeepAsrani commented on src/pro/blocks/modal/block.json:
```diff
@@ -0,0 +1,150 @@
+		"maxWidth": {
+			"type": ["number", "string"]
+		},
+		"trigger": {
```
Why do we have these unused attributes from the Popup block?

Soare-Robert-Daniel replied:
I kept it for compatibility. Should I also transform from Popup to Modal and vice versa?

HardeepAsrani replied:
Yea, you can add transformation but just remove the trigger as both are separate blocks and this attribute doesn't do anything in modal.

HardeepAsrani commented on src/blocks/blocks/modal/block.json:
We only need the attributes in the free version if this exists. Does the editor remove the attributes if you switch to free without having the attrs specified?

Soare-Robert-Daniel replied:
If the both `block.json` have different attributes, the missing one will be auto-cleaned by Gutenberg.

HardeepAsrani replied:
Have you tested it? If yes then sure we can keep it like this.

Soare-Robert-Daniel replied:
yeah. I did

HardeepAsrani commented on src/pro/blocks/modal/edit.js:
```diff
@@ -0,0 +1,247 @@
+
+import classnames from 'classnames';
+
+import { get } from 'lodash';
+
+import { merge } from 'lodash';
```
```suggestion
import {
	get,
	merge
} from 'lodash';
```

HardeepAsrani reviewed (changes requested):
Looks all good, left a couple of small things but overall it works nicely.

ineagu commented:
> @ineagu @HardeepAsrani When a Free user uses the Modal block which Pro, it will see this:
> 
> ![image](https://private-user-images.githubusercontent.com/17597852/352066787-b939400d-9d4f-4897-9a05-b6323403b4e3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwMjAzNDcsIm5iZiI6MTcyMzAyMDA0NywicGF0aCI6Ii8xNzU5Nzg1Mi8zNTIwNjY3ODctYjkzOTQwMGQtOWQ0Zi00ODk3LTlhMDUtYjYzMjM0MDNiNGUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA3VDA4NDA0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmODM5NTEzMWNmNDc5ODk4MTdmNTI0YmU2MzUxYjZiOTQwM2ZiMTYwYzY0YjFmZWFhMzUxYWJkZDFlYjQxMDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qmwRgSDxvRmyHxV63QxwCbw1yo03fx3MW3l4clfh5HY)
> 
> Two questions:
> 
> * Is the upsell message good?
> * Should we use a different icon for the block to be different from the popup (the popup icon is used in the image)?

this message looks good to me, we can go with the same icon for now, I couldn't find a good one.

rodica-andronache commented:
@Soare-Robert-Daniel tested and everything works well, except for a small thing with the close button.

If I choose a border color and position: outside, everything looks good in the editor.
![Annotation on 2024-08-07 at 11-46-16](https://github.com/user-attachments/assets/940cb891-28a9-4697-bfef-afa0a67de98a)

![Screenshot on 2024-08-07 at 12-00-27](https://github.com/user-attachments/assets/a4226684-bb3f-4bf6-b4d4-04d5917cdcf9)

Using Neve, the close button color is appearing ok, but the position of the close button is not appearing outside
![Screenshot on 2024-08-07 at 12-00-18](https://github.com/user-attachments/assets/cd20c781-a567-408d-bc2c-19769eeaa1ef)

Using the default theme 2024, neither the button color or position applies
![Screenshot on 2024-08-07 at 11-46-31](https://github.com/user-attachments/assets/734ff9ed-fe92-4e9e-a10c-0c08f800b498)

Soare-Robert-Daniel commented:
@rodica-andronache, the button position should work now, fixed the malformed class name.

Regarding the button color, in the new version of the WP Theme, some CSS properties are no longer added by the core, so I added them to our files.

rodica-andronache commented:
@Soare-Robert-Daniel for me, it's still not working on frontend, either with Neve or 2024, the outside position of the close button is not working. 
![Screenshot on 2024-08-07 at 17-10-14](https://github.com/user-attachments/assets/f68d5a9e-491f-4285-bfb5-3f107aa6cd2e)
Plus, now in the backend, the image is also outside the container when you choose Outside position, for the close button.
![Screenshot on 2024-08-07 at 17-11-18](https://github.com/user-attachments/assets/b7411f28-99e3-44c1-92bc-f8bc92d8c5f5)

Also, I've noticed the block is missing the PRO tag, when using just the free version
![Annotation on 2024-08-07 at 17-05-51](https://github.com/user-attachments/assets/8dfb4343-341b-453d-b29c-cb3b8fe8ba30)

Soare-Robert-Daniel commented:
The click outside should have been fixed.


https://github.com/user-attachments/assets/56487f1c-6edc-4037-a27f-f458c8a969bf

---

Regarding the out-the-frame window. It seems that WP changed some CSS, and our fixed offset of `40px` does not seem to work nicely for the new version.

This is problematic for some blocks like Core Image, which like to not respect their frame nicely. To avoid breaking compatibility with older versions, I made a trick in which I put the default template inside a stack with a top margin of `40px` to counterbalance. 

---

The PRO labels in dashboard is not fixed

rodica-andronache commented:
@Soare-Robert-Daniel the issues are fixed now 👍 
I have just a small question: Is it expected to not be able to have different heights for Desktop/Mobile? Meaning, to have **Fit Content** on Desktop and a **Custom** height on mobile https://vertis.d.pr/i/8SxCxV

Soare-Robert-Daniel commented:
@rodica-andronache, I think yes, it is the same behavior as Popup. I'm not sure if @HardeepAsrani remembers the reason for this.
